README.pod
lib/Module/Advisor.pod

Changes
Revision history for Perl extension Module::Advisor

{{$NEXT}}

0.17 2014-02-20T08:00:49Z

    - Notes about Plack::Middleware::AxsLog

0.16 2013-09-30T01:25:54Z

    - Added script, basically the SYNOPSIS in a script
      (Neil Bowers)

0.15 2013-09-29T01:50:45Z

    - Notes Mouse's memory leak bug

0.14 2013-04-30T22:51:05Z

    - Notes autobox 2.78

0.13 2013-04-27T14:07:44Z

    - Notes Module::Install::XSUtil

0.12 2013-03-05T15:12:56

    - Notes about Encode@2.49 memory leak fix

0.11 2013-02-08T11:42:08

    - use Module::Metadata and CPAN::Meta::Requirements
      (miyagawa++)

0.10 2013-02-08T10:55:00

    - Mouse 1.04 has memory leaks

0.09 2013-01-22T12:33:08

    - added ExclusiveLock::Guard 0.04

0.08 2013-01-20T19:46:33

    - better documentation

0.07 2013-01-20T19:14:17

    - renamed to Module::Advisor from Acme::Module::Checker.
    - note about Proc::daemon

0.06 2012-12-04

    - HTML::FillInForm::Lite

0.05 2012-12-04

    - note xslate bug

0.04 2012-12-03

    - notes about Amon2::DBI

0.03 2012-10-19

    - more.

0.02 2012-10-14

    - note on older xslate

0.01 2012-09-27 09:46:04
    - original version (called Acme::Module::Checker)



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Module-Advisor',
    module_name     => 'Module::Advisor',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


author/docgen.pl
#!/usr/bin/env perl
use strict;
use warnings;
use utf8;
use 5.010000;
use autodie;
use Text::MicroTemplate qw/render_mt/;
use lib 'lib';
use Module::Advisor;

our %pod_escape_table = ('>' => 'E<gt>', '<' => 'E<lt>');

sub escape_pod {
    my ($variable) = @_;

    my $source = qq{
        do{
            $variable =~ s/([><])/\$::pod_escape_table{\$1}/ge;
            $variable;
        }
    };
    $source =~ s/\n//g; # to keep line numbers
    return $source;
}

my $tmpl = join '', <DATA>;
my $render = Text::MicroTemplate->new(template => $tmpl,
                                      escape_func => \&escape_pod);
my $result = $render->build->();
print $result;

__DATA__

=encoding utf8

=head1 NAME

Module::Advisor - check a modules you are installed

=head1 SYNOPSIS

    use Module::Advisor;
    Module::Advisor->new()->check();

=head1 DESCRIPTION

Module::Advisor checks a modules you are installed, and notice if:

=over 4

=item There is a module, have known bugs.

=item There is a module, have optional XS module for better performance.

=item Your are using broken version of CPAN module.

=back

=head1 RULES

Here is a rules to check the modules.

=head2 Modules have security issues.

=over 4

? for my $module (@Module::Advisor::SECURITY) {
=item <?= $module->[0] ?> <?= $module->[1] ?>

<?= $module->[2] ?>

? }

=back

=head2 Modules have performance improvements

=over 4

? for my $module (@Module::Advisor::PERFORMANCE) {

=item <?= $module->[0] ?> <?= $module->[1] ?>

<?= $module->[2] ?>

? }

=back

=head2 Modules have bugs

=over 4

? for my $module (@Module::Advisor::BUG) {

=item <?= $module->[0] ?> <?= $module->[1] ?>

<?= $module->[2] ?>

? }

=back

=head2 The version of this module was broken

=over 4

? for my $module (@Module::Advisor::BROKEN) {

=item <?= $module->[0] ?> <?= $module->[1] ?> has bug.

<?= $module->[2] ?>

? }

=back

=head2 Recommended XS module for better performance

=over 4

? for my $module (@Module::Advisor::XS) {

=item <?= $module->[1] ?> is recommended

If you are using <?= $module->[0] ?>.

? }

=back

=head2 Recommended version to enable good feature

=over 4

? for my $module (@Module::Advisor::FEATURE) {

=item <?= $module->[0] ?> <?= $module->[1] ?> does not have

<?= $module->[2] ?>

? }

=back

=head2 Recommend modules when using ...

=over 4

? for my $module (@Module::Advisor::OPTIONAL_MODULES) {

=item I recommend to install <?= $module->[1] ?>, if you are using <?= $module->[0] ?>.

<?= $module->[2] ?>

? }

=back

=head1 AUTHOR

Tokuhiro Matsuno E<lt>tokuhirom AAJKLFJEF@ GMAIL COME<gt>

=head1 SEE ALSO

=head1 LICENSE

Copyright (C) Tokuhiro Matsuno

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Acme/Module/Checker.pm
package Acme::Module::Checker;
use strict;
use warnings;
use 5.008008;
our $VERSION = '0.06';
use version;
use ExtUtils::MakeMaker;
use Carp;

my @SECURITY = (
    ['Digest', '1.17', '<1.16 have a security issue. which could lead to the injection of arbitrary Perl code'],
    ['Encode' => '2.44', 'heap overflow'],
);

my @PERFORMANCE = (
    ['UNIVERSAL::require', 0.11, '0.11+ is 400% faster'],
);

my @BUG = (
    ['Plack' => '0.9982', 'sanity check to remove newlines from headers'],
    ['Time::Piece' => '1.16', '<1.15 have timezone related issue'],
    ['DBD::SQLite' => '1.20', 'a lot of bugs.'],
    ['Text::Xslate' => '1.0011', '&apos; bug.'],
    ['Text::Xslate' => '1.5021', 'segv in "render" recursion call'],
    ['Text::Xslate' => '1.6001', '<1.6001 possibly memory leaks on VM stack frames. see https://github.com/xslate/p5-Text-Xslate/issues/71'],
    ['Furl' => '0.39', 'unexpected eof in reading chunked body. It makes busy loop.'],
    ['AnyEvent::MPRPC' => '0.15', 'switch to Data::MessagePack::Stream'],
    ['Data::MessagePack' => '0.46', 'fixed unpacking issue on big-endian system.'],
    ['Data::MessagePack' => '0.39', 'packing float numbers fails on some cases'],
    ['FCGI::Client' => '0.06', 'fixed large packet issue'],
    ['Starlet' => 0.12, 'fix infinite loop when connection is closed while receiving response content'],
    ['Starman' => 0.2014, '$res->[1] is broken after output (This is actualized with Plack::Middleware::AccessLog::Timed) https://github.com/miyagawa/Starman/pull/31'],
    ['Starman' => 0.1006, 'Fixed 100% CPU loop when an unexpected EOF happens'],
    ['Twiggy', '0.1000', 'busy loop'],
    ['Teng', '0.14', 'fixed deflate bug.'],
    ['DBIx::Skinny', 0.0742, 'txn_scope bug fixed'],
    ['DBIx::TransactionManager', 1.11, 'not execute begin_work at AutoCommit=0.'],
    ['HTTP::MobileAgent' => '0.36', 'new x-up-devcap-multimedia(StandAloneGPS) support'],
    ['HTTP::MobileAgent' => '0.35', 'Updated $HTMLVerMap and $GPSModelsRe in DoCoMo.pm'],
    ['Encode::JP::Mobile' => '0.25', 'resolved FULLWIDTH TILDE issue, etc.'],
    ['Template' => '2.15', 'uri filter does not works properly https://rt.cpan.org/Public/Bug/Display.html?id=19593'],
    ['HTML::FillInForm::Lite' => '1.11', 'HTML5 style tags support'],
    ['Proc::Daemon' => '0.12', 'Init() did not close all filehandles reliably in some cases.'],
);

my @BROKEN = (
    ['Amon2::DBI' => '0.31', 'transaction management bug'],
    ['Math::Random::MT' => '1.15', 'rand() took no notice of argument RT #78200'],
    ['Module::Install' => '1.04', 'Broken, http://weblog.bulknews.net/post/33907905561/do-not-ship-modules-with-module-install-1-04'],
);

my @XS = (
    ['JSON' => 'JSON::XS'],
    # ['PPI' => 'PPI::XS'], # I think PPI::XS is outdated.
    ['Plack' => 'HTTP::Parser::XS'],
);
if ($^O eq 'linux') {
    push @XS, ['Filesys::Notify::Simple', 'Linux::Inotify2'];
}

my @FEATURE = (
    ['Amon2' => '3.29', 'JSON hijacking detection.'],
    ['Log::Minimal' => '0.10', 'LM_COLOR'],
    ['Log::Minimal' => '0.08', 'colourful logging'],
    ['Log::Minimal' => '0.03', 'ddf'],
    ['Proclet' => 0.12, 'Proclet::Declare'],
    ['DBI', '1.614' => 'AutoInactiveDestroy'],
);

my @OPTIONAL_MODULES = (
    ['LWP', 'LWP::Protocol::https', 'Need to support https'],
);

# ref. Module::Version
sub get_version {
    my $module = shift or croak 'Must get a module name';
    my $file = MM->_installed_file_for_module($module);

    $file || return;

    return MM->parse_version($file);
}

sub debugf {
    my $self = shift;
    print "# @_\n" if $self->{verbose};
}

sub new {
    my $class = shift;
    my %args = @_==1 ? %{$_[0]} : @_;
    bless {%args}, $class;
}

sub check {
    my ($self) = @_;
    my $failed = 0;
    $failed += $self->check_security();
    $failed += $self->check_performance();
    $failed += $self->check_bugs();
    $failed += $self->check_bugs_on_specific_version();
    $failed += $self->check_xs_installed();
    $failed += $self->check_feature();
    return $failed;
}

sub _check_issue {
    my ($self, $patterns, $type_name) = @_;
    my $failed = 0;
    for my $row (@$patterns) {
        my $ver = get_version($row->[0]);
        if (defined $ver) {
            $self->debugf("$row->[0] $ver");
            if (version->new($ver) < version->new($row->[1])) {
                printf "[$type_name] %s %s<%s: %s\n", $row->[0], $ver, $row->[1], $row->[2];
                $failed++;
            }
        } else {
            $self->debugf("$row->[0] is not found");
        }
    }
    return $failed;
}

sub check_security {
    my $self = shift;
    $self->_check_issue(\@SECURITY, 'SECURITY');
}

sub check_performance {
    my $self = shift;
    $self->_check_issue(\@PERFORMANCE, 'PERFORMANCE');
}

sub check_bugs {
    my $self = shift;
    $self->_check_issue(\@BUG, 'BUG');
}

sub check_bugs_on_specific_version {
    my ($self) = @_;
    my $failed = 0;
    for my $row (@BROKEN) {
        my $ver = get_version($row->[0]);
        if (defined $ver) {
            $self->debugf("$row->[0] $ver");
            if (version->new($ver) == version->new($row->[1])) {
                printf "[BROKEN] %s %s: %s\n", $row->[0], $ver, $row->[2];
                $failed++;
            }
        } else {
            $self->debugf("$row->[0] is not found");
        }
    }
}

sub check_xs_installed {
    my ($self) = @_;
    my $failed = 0;
    for my $row (@XS) {
        my $pmver = get_version($row->[0]);
        if (defined $pmver) {
            my $xsver = get_version($row->[1]);
            if (defined $xsver) {
                $self->debugf("$row->[1] is also installed");
            } else {
                printf "[XS] %s is need to install for better performance for %s\n", $row->[1], $row->[0];
            }
        } else {
            $self->debugf("$row->[0] is not installed");
        }
    }
}

sub _check_optional {
    my ($self, $rules, $type_name) = @_;
    my $failed = 0;
    for my $row (@$rules) {
        my $pmver = get_version($row->[0]);
        if (defined $pmver) {
            my $xsver = get_version($row->[1]);
            if (defined $xsver) {
                $self->debugf("$row->[1] is also installed");
            } else {
                printf "[$type_name] %s => %s: %s\n", $row->[1], $row->[0], $row->[2];
            }
        } else {
            $self->debugf("$row->[0] is not installed");
        }
    }
}

sub check_feature {
    my $self = shift;
    $self->_check_optional(\@OPTIONAL_MODULES, 'OPTIONAL_MODULES');
}

if (not defined caller(0)) {
    my $checker = __PACKAGE__->new();
    $checker->{verbose}++ if $ENV{DEBUG};
    $checker->check() or print "OK\n";
}

1;
__END__

=encoding utf8

=head1 NAME

Acme::Module::Checker - check a modules you are installed

=head1 SYNOPSIS

    use Acme::Module::Checker;
    Acme::Module::Checker->new()->check();

=head1 DESCRIPTION

Acme::Module::Checker checks a modules you are installed, in tokuhirom's rule.

=head1 AUTHOR

Tokuhiro Matsuno E<lt>tokuhirom AAJKLFJEF@ GMAIL COME<gt>

=head1 SEE ALSO

=head1 LICENSE

Copyright (C) Tokuhiro Matsuno

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Module/Advisor.pm
package Module::Advisor;
use strict;
use warnings;
use 5.008008;
our $VERSION = '0.16';
use Module::Metadata;
use CPAN::Meta::Requirements;
use Carp;

our @SECURITY = (
    ['Digest' => '< 1.17', 'have a security issue. which could lead to the injection of arbitrary Perl code'],
    ['Encode' => '< 2.44', 'heap overflow'],
    ['Encode' => '< 2.49', 'memory leaks'],
);

our @PERFORMANCE = (
    ['UNIVERSAL::require', '< 0.11', 'is 400% faster'],
);

our @BUG = (
    ['Plack' => '< 0.9982', 'sanity check to remove newlines from headers'],
    ['Time::Piece' => '< 1.16', 'have timezone related issue'],
    ['DBD::SQLite' => '< 1.20', 'a lot of bugs.'],
    ['Mouse' => '>= 1.07, < 1.12', 'Memory leaks in trigger'],
    ['Text::Xslate' => '< 1.0011', '&apos; bug.'],
    ['Text::Xslate' => '< 1.5021', 'segv in "render" recursion call'],
    ['Text::Xslate' => '< 1.6001', 'possibly memory leaks on VM stack frames. see https://github.com/xslate/p5-Text-Xslate/issues/71'],
    ['Text::Xslate' => '< 2.0005', 'Nested WRAPPER broken https://github.com/xslate/p5-Text-Xslate/issues/79'],
    ['Furl' => '< 0.39', 'unexpected eof in reading chunked body. It makes busy loop.'],
    ['AnyEvent::MPRPC' => '< 0.15', 'switch to Data::MessagePack::Stream'],
    ['Data::MessagePack' => '< 0.46', 'fixed unpacking issue on big-endian system.'],
    ['Data::MessagePack' => '< 0.39', 'packing float numbers fails on some cases'],
    ['FCGI::Client' => '< 0.06', 'fixed large packet issue'],
    ['Starlet' => '< 0.12', 'fix infinite loop when connection is closed while receiving response content'],
    ['Starman' => '< 0.2014', '$res->[1] is broken after output (This is actualized with Plack::Middleware::AccessLog::Timed) https://github.com/miyagawa/Starman/pull/31'],
    ['Starman' => '< 0.1006', 'Fixed 100% CPU loop when an unexpected EOF happens'],
    ['Twiggy' => '< 0.1000', 'busy loop'],
    ['Teng', '< 0.14', 'fixed deflate bug.'],
    ['DBIx::Skinny', '< 0.0742', 'txn_scope bug fixed'],
    ['DBIx::TransactionManager', '< 1.11', 'not execute begin_work at AutoCommit=0.'],
    ['HTTP::MobileAgent' => '< 0.36', 'new x-up-devcap-multimedia(StandAloneGPS) support'],
    ['HTTP::MobileAgent' => '< 0.35', 'Updated $HTMLVerMap and $GPSModelsRe in DoCoMo.pm'],
    ['Encode::JP::Mobile' => '< 0.25', 'resolved FULLWIDTH TILDE issue, etc.'],
    ['Template' => '< 2.15', 'uri filter does not works properly https://rt.cpan.org/Public/Bug/Display.html?id=19593'],
    ['HTML::FillInForm::Lite' => '< 1.11', 'HTML5 style tags support'],
    ['Proc::Daemon' => '< 0.12', 'Init() did not close all filehandles reliably in some cases.'],
    ['ExclusiveLock::Guard' => '< 0.04', 'change of the file stat timing (measures under high load)'],
    ['autobox', '< 2.78', 'segv in END block https://rt.cpan.org/Ticket/Display.html?id=80400'],

    # Broken in specific versions
    ['Amon2::DBI' => '== 0.31', 'transaction management bug'],
    ['Math::Random::MT' => '== 1.15', 'rand() took no notice of argument RT #78200'],
    ['Module::Install' => '== 1.04', 'Broken, http://weblog.bulknews.net/post/33907905561/do-not-ship-modules-with-module-install-1-04'],
    ['Mouse' => '== 1.04', 'Broken, http://d.hatena.ne.jp/gfx/20130208/1360283357'],
    ['Plack::Middleware::AxsLog' => '== 0.20', 'Missing \n'],
);

our @XS = (
    ['JSON' => 'JSON::XS'],
    # ['PPI' => 'PPI::XS'], # I think PPI::XS is outdated.
    ['Plack' => 'HTTP::Parser::XS'],
);
if ($^O eq 'linux') {
    push @XS, ['Filesys::Notify::Simple', 'Linux::Inotify2'];
}

our @FEATURE = (
    ['Amon2' => '< 3.29', 'JSON hijacking detection.'],
    ['Log::Minimal' => '< 0.10', 'LM_COLOR'],
    ['Log::Minimal' => '< 0.08', 'colourful logging'],
    ['Log::Minimal' => '< 0.03', 'ddf'],
    ['Proclet' => '< 0.12', 'Proclet::Declare'],
    ['DBI', '< 1.614' => 'AutoInactiveDestroy'],
    ['Module::Install::XSUtil', '< 0.44' => 'PUREPERL_ONLY=1'],
);

our @OPTIONAL_MODULES = (
    ['LWP', 'LWP::Protocol::https', 'Need to support https'],
);

# ref. Module::Version
sub get_version {
    my $module = shift or croak 'Must get a module name';
    my $metadata = Module::Metadata->new_from_module($module);
    $metadata ? $metadata->version : undef;
}

sub debugf {
    my $self = shift;
    print "# @_\n" if $self->{verbose};
}

sub new {
    my $class = shift;
    my %args = @_==1 ? %{$_[0]} : @_;
    bless {%args}, $class;
}

sub check {
    my ($self) = @_;
    my $failed = 0;
    $failed += $self->check_security();
    $failed += $self->check_performance();
    $failed += $self->check_bugs();
    $failed += $self->check_xs_installed();
    $failed += $self->check_feature();
    return $failed;
}

sub matches_version {
    my($self, $module, $spec, $version) = @_;

    my $requirements = CPAN::Meta::Requirements->new;
    $requirements->add_string_requirement($module, $spec);

    $requirements->accepts_module($module, $version);
}

sub _check_issue {
    my ($self, $patterns, $type_name) = @_;
    my $failed = 0;
    for my $row (@$patterns) {
        my $ver = get_version($row->[0]);
        if (defined $ver) {
            $self->debugf("$row->[0] $ver");
            if ($self->matches_version($row->[0], $row->[1], $ver)) {
                printf "[$type_name] %s %s %s: %s\n", $row->[0], $ver, $row->[1], $row->[2];
                $failed++;
            }
        } else {
            $self->debugf("$row->[0] is not found");
        }
    }
    return $failed;
}

sub check_security {
    my $self = shift;
    $self->_check_issue(\@SECURITY, 'SECURITY');
}

sub check_performance {
    my $self = shift;
    $self->_check_issue(\@PERFORMANCE, 'PERFORMANCE');
}

sub check_bugs {
    my $self = shift;
    $self->_check_issue(\@BUG, 'BUG');
}

sub check_xs_installed {
    my ($self) = @_;
    my $failed = 0;
    for my $row (@XS) {
        my $pmver = get_version($row->[0]);
        if (defined $pmver) {
            my $xsver = get_version($row->[1]);
            if (defined $xsver) {
                $self->debugf("$row->[1] is also installed");
            } else {
                printf "[XS] %s is need to install for better performance for %s\n", $row->[1], $row->[0];
            }
        } else {
            $self->debugf("$row->[0] is not installed");
        }
    }
}

sub _check_optional {
    my ($self, $rules, $type_name) = @_;
    my $failed = 0;
    for my $row (@$rules) {
        my $pmver = get_version($row->[0]);
        if (defined $pmver) {
            my $xsver = get_version($row->[1]);
            if (defined $xsver) {
                $self->debugf("$row->[1] is also installed");
            } else {
                printf "[$type_name] %s => %s: %s\n", $row->[1], $row->[0], $row->[2];
            }
        } else {
            $self->debugf("$row->[0] is not installed");
        }
    }
}

sub check_feature {
    my $self = shift;
    $self->_check_optional(\@OPTIONAL_MODULES, 'OPTIONAL_MODULES');
}

if (not defined caller(0)) {
    my $checker = __PACKAGE__->new();
    $checker->{verbose}++ if $ENV{DEBUG};
    $checker->check() or print "OK\n";
}

1;
